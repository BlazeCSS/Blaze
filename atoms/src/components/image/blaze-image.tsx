import { Component, Prop } from '@stencil/core';

@Component({
  tag: 'blaze-image'
})
export class Image {

  @Prop() src: string;
  @Prop() alt: string;

  // Placeholder settings
  @Prop() width: number;
  @Prop() height: number;
  @Prop() photo: string;        // specific photo
  @Prop() user: string;         // random photo from a user
  @Prop() likes: boolean;       // random photo from a user's likes
  @Prop() collection: string;   // random photo from a collection
  @Prop() filter: string = '';  // filter based on search term

  generateSrc() {
    if (this.src) return this.src;

    let size = '';
    if (this.width && this.height) {
      size = `${this.width}x${this.height}/`
    }

    let filter = '';
    if (this.filter) {
      filter = `?${this.filter}`;
    }

    if (this.photo) {
      return `https://source.unsplash.com/${this.photo}/${size}`;
    }

    if (this.user) {
      return this.likes ?
        `https://source.unsplash.com/user/${this.user}/likes/${size}${filter}` :
        `https://source.unsplash.com/user/${this.user}/${size}${filter}`
    }

    if (this.collection) {
      return `https://source.unsplash.com/collection/${this.collection}/${size}${filter}`
    }

    return `https://source.unsplash.com/random/${size}${filter}`;
  }

  render() {
    const src = this.generateSrc();

    return (
      <img class="o-image" src={src} alt={this.alt} />
    );
  }
}
