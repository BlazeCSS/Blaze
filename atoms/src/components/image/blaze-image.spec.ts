import { flush, render } from '@stencil/core/testing';
import { Image } from './blaze-image';

describe('Image', () => {
  it('should build', () => {
    expect(new Image()).toBeTruthy();
  });

  let element;

  const snapIt = (name, html) => {
    it(name, async () => {
      element = await render({
        components: [Image],
        html
      });
      await flush(element);

      expect(element).toMatchSnapshot();
    });
  };

  snapIt(
    'renders a standard image with src and alt attributes',
    '<blaze-image src="https://www.blazeui.com/assets/img/blaze.png" alt="blaze logo"></blaze-image>'
  );

  snapIt(
    'falls back to a random placeholder image',
    '<blaze-image></blaze-image>'
  );

  snapIt(
    'falls back to a random placeholder image with a filter',
    '<blaze-image filter="space"></blaze-image>'
  );

  snapIt(
    'width and height',
    '<blaze-image width="1024" height="768"></blaze-image>'
  );

  snapIt(
    'with specific photo id',
    '<blaze-image photo="test-photo-id"></blaze-image>'
  );

  snapIt(
    'with specific photo id and width and height',
    '<blaze-image photo="test-photo-id" width="1024" height="768"></blaze-image>'
  );

  snapIt(
    'from a specific user',
    '<blaze-image user="test-user"></blaze-image>'
  );

  snapIt(
    'from a specific user with width and height',
    '<blaze-image user="test-user" width="1024" height="768"></blaze-image>'
  );

  snapIt(
    'from a specific user with width and height and an additional filter',
    '<blaze-image user="test-user" width="1024" height="768" filter="plane"></blaze-image>'
  );

  snapIt(
    'from the list of a specific user\'s like photos with additional attributes',
    '<blaze-image user="test-user" width="1024" height="768" filter="mountains" likes></blaze-image>'
  );
});
