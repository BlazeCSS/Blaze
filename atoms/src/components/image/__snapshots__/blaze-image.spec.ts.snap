// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Image falls back to a random placeholder image 1`] = `
<blaze-image
  class="hydrated"
>
  <!---->
  <img
    class="o-image"
    src="https://source.unsplash.com/random/"
  />
</blaze-image>
`;

exports[`Image falls back to a random placeholder image with a filter 1`] = `
<blaze-image
  class="hydrated"
  filter="space"
>
  <!---->
  <img
    class="o-image"
    src="https://source.unsplash.com/random/?space"
  />
</blaze-image>
`;

exports[`Image from a specific user 1`] = `
<blaze-image
  class="hydrated"
  user="test-user"
>
  <!---->
  <img
    class="o-image"
    src="https://source.unsplash.com/user/test-user/"
  />
</blaze-image>
`;

exports[`Image from a specific user with width and height 1`] = `
<blaze-image
  class="hydrated"
  height="768"
  user="test-user"
  width="1024"
>
  <!---->
  <img
    class="o-image"
    src="https://source.unsplash.com/user/test-user/1024x768/"
  />
</blaze-image>
`;

exports[`Image from a specific user with width and height and an additional filter 1`] = `
<blaze-image
  class="hydrated"
  filter="plane"
  height="768"
  user="test-user"
  width="1024"
>
  <!---->
  <img
    class="o-image"
    src="https://source.unsplash.com/user/test-user/1024x768/?plane"
  />
</blaze-image>
`;

exports[`Image from the list of a specific user's like photos with additional attributes 1`] = `
<blaze-image
  class="hydrated"
  filter="mountains"
  height="768"
  likes=""
  user="test-user"
  width="1024"
>
  <!---->
  <img
    class="o-image"
    src="https://source.unsplash.com/user/test-user/likes/1024x768/?mountains"
  />
</blaze-image>
`;

exports[`Image renders a standard image with src and alt attributes 1`] = `
<blaze-image
  alt="blaze logo"
  class="hydrated"
  src="https://www.blazeui.com/assets/img/blaze.png"
>
  <!---->
  <img
    alt="blaze logo"
    class="o-image"
    src="https://www.blazeui.com/assets/img/blaze.png"
  />
</blaze-image>
`;

exports[`Image width and height 1`] = `
<blaze-image
  class="hydrated"
  height="768"
  width="1024"
>
  <!---->
  <img
    class="o-image"
    src="https://source.unsplash.com/random/1024x768/"
  />
</blaze-image>
`;

exports[`Image with specific photo id 1`] = `
<blaze-image
  class="hydrated"
  photo="test-photo-id"
>
  <!---->
  <img
    class="o-image"
    src="https://source.unsplash.com/test-photo-id/"
  />
</blaze-image>
`;

exports[`Image with specific photo id and width and height 1`] = `
<blaze-image
  class="hydrated"
  height="768"
  photo="test-photo-id"
  width="1024"
>
  <!---->
  <img
    class="o-image"
    src="https://source.unsplash.com/test-photo-id/1024x768/"
  />
</blaze-image>
`;
